lib/ui/painting/image_decoder.cc

```diff
@@ -42,11 +42,22 @@ static double AspectRatio(const SkISize& size) {
 // intrinsic dimensions of the image.
 static SkISize GetResizedDimensions(SkISize current_size,
                                     std::optional<uint32_t> target_width,
-                                    std::optional<uint32_t> target_height) {
+                                    std::optional<uint32_t> target_height,
+                                    ImageUpscalingMode image_upscaling) {
   if (current_size.isEmpty()) {
     return SkISize::MakeEmpty();
   }
 
+  if (image_upscaling == ImageUpscalingMode::kNotAllowed) {
+    if (target_width) {
```

Comment:
This is a good point.

My grand plan is to follow this up with another PR to add something to control how aspect ratios are or are not preserved. I was originally thinking that this could be separated out from that, but I think you're right that it can't. 

I'll take a quick look at how much more complicated this PR gets to add that extra parameter, otherwise I'll look to clarify the docs.

Reply:
Actually, if a user wants that case they have to just set the `allowUpscaling` bit to true right?

But yes, we should document and test this.

Reply:
Yeah, we have sort of tacked on stuff to this original simple API in bits and pieces as we went along. This has understandably become a bit unwieldy.

> I'll take a quick look at how much more complicated this PR gets to add that extra parameter...

Instead of doing all the design work to further update this API with additional options, I am fine if we just say "this will/won't preserve aspect ratio if both dimensions are specified" and add a test for this. As I said, my comment was more about the undefined behavior. This is still an improvement for a vast majority of the cases.

In the long run, an API that works with image descriptors without decompression is the way to go. But that is a lot more work to write and migrate to.

Reply:
> if a user wants that case they have to just set the allowUpscaling bit to true right?

Yeah, there is a workaround. Let's just document and test the default. My only concern still is that since the default has changed, earlier, they didn't need to specify the argument but now they have to.

How about just preserving the aspect ratio while scaling down if both width and height are specified? That ought to be really straightforward.

Reply:
Ok, the more I look at this, the more I think it's just a doc and test thing. Whether or how we add other aspect ratio parameters here will still be relevant to this case.

Reply:
Sounds good.